File: backend/src/main/java/api/service/BlogService.java
package api.service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.domain.Sort.Direction;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import api.model.blog.Blog;
import api.model.blog.BlogRequest;
import api.model.blog.BlogResponse;
import api.model.blog.ChildrenResponse;
import api.model.like.Like;
import api.model.user.User;
import api.repository.BlogRepository;
import api.repository.LikesRepository;
import api.repository.ReportRepository;
import api.repository.UserRepository;

/**
 * Service for handling blog-related business logic.
 * Provides methods for creating, retrieving, updating, deleting, and interacting with blogs.
 */
@Service
public class BlogService {

    private final BlogRepository blogRepository;
    private final LikesRepository likesRepository;
    private final UserRepository userRepository;
    private final ReportRepository reportRepository;

    private final NotificationService notificationService;
    private static final String ERROR_USER_NOT_FOUND = "User not found with ID: %d";

    public BlogService(BlogRepository blogRepository, UserRepository userRepository,
            NotificationService notificationService, CloudinaryService cloudinaryService,
            ReportRepository reportRepository, LikesRepository likesRepository) {
        this.blogRepository = blogRepository;
        this.userRepository = userRepository;
        this.notificationService = notificationService;
        this.reportRepository = reportRepository;
        this.likesRepository = likesRepository;
    }

    /**
     * Retrieves a paginated list of all blogs.
     * @param cursor The pagination cursor.
     * @return A list of blogs.
     */
    public List<BlogResponse> getBlogs(long cursor) {
        User user = getAuthenticatedUser();
        Pageable pageable = PageRequest.of(0, 10, Direction.DESC, "id");
        if (cursor == 0) {
            cursor = Long.MAX_VALUE;
        }
        return this.blogRepository
                .findByParentIsNullAndIdLessThan(cursor, pageable)
                .stream()
                .map(b -> new BlogResponse(b, user.getId()))
                .toList();
    }

    /**
     * Retrieves the total count of all blogs.
     * @return The total blog count.
     */
    public long getBlogsCount() {
        return this.blogRepository.count();
    }

    /**
     * Retrieves blogs from the user's network.
     * @param cursor The pagination cursor.
     * @return A list of blogs from the user's network.
     */
    public List<BlogResponse> getBlogsNetworks(long cursor) {
        User user = getAuthenticatedUser();
        Pageable pageable = PageRequest.of(0, 10, Direction.DESC, "id");
        if (cursor == 0) {
            cursor = Long.MAX_VALUE;
        }
        return this.blogRepository
                .findSubscribedUsersBlogs(user.getId(), cursor, pageable)
                .stream()
                .map(b -> new BlogResponse(b, user.getId()))
                .toList();
    }

    /**
     * Retrieves a list of blogs created by a specific user.
     * @param userId The ID of the user.
     * @param cursor The pagination cursor.
     * @return A list of blogs created by the user.
     */
    public List<BlogResponse> getBlogsByUser(long userId, long cursor) {
        User user = getAuthenticatedUser();
        Pageable pageable = PageRequest.of(0, 10, Direction.DESC, "id");
        if (cursor == 0) {
            cursor = Long.MAX_VALUE;
        }
        return this.blogRepository
                .findByUserIdAndParentIsNullAndIdLessThanAndHiddenFalse(userId, cursor, pageable)
                .stream()
                .map(b -> new BlogResponse(b, user.getId()))
                .toList();
    }

    /**
     * Saves a new blog post.
     * @param blogRequest The request body containing blog details.
     * @return The saved blog post.
     */
    @Transactional
    public BlogResponse saveBlog(BlogRequest blogRequest) {
        User user = getAuthenticatedUser();
        // validateMediaType(blogRequest.mediaType());
        Blog blog = convertToEntity(blogRequest);
        blog.setUser(user);
        blog.setCreatedAt(LocalDateTime.now());
        Blog savedBlog = this.blogRepository.save(blog);
        // save notifaction
        if (blogRequest.parent() == null) {
            this.notificationService.saveNotification(savedBlog);
        }
        return new BlogResponse(savedBlog, user.getId());
    }

    /**
     * Retrieves the currently authenticated user.
     * @return The authenticated user.
     * @throws ResponseStatusException if the user is not found.
     */
    public User getAuthenticatedUser() {
        User user = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        return userRepository.findById(user.getId())
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.NOT_FOUND, String.format(ERROR_USER_NOT_FOUND, user.getId())));
    }

    /**
     * Retrieves a single blog post by its ID.
     * @param blogId The ID of the blog to retrieve.
     * @return The requested blog post.
     */
    public BlogResponse getBlog(long blogId) {
        User user = getAuthenticatedUser();
        return this.blogRepository
                .findById(blogId)
                .map(b -> new BlogResponse(b, user.getId()))
                .get();
    }

    /**
     * Deletes a blog post by its ID.
     * @param blogId The ID of the blog to delete.
     */
    @Transactional
    public void deleteBlog(long blogId) {
        Blog blog = this.blogRepository.findById(blogId).get();
        this.reportRepository.findByBlogId(blog.getId()).forEach(report -> {
            report.setBlogId(0L);
            this.reportRepository.save(report);
        });
        this.blogRepository.deleteById(blog.getId());
    }

    /**
     * Retrieves the children (comments) of a blog post.
     * @param blogId The ID of the parent blog.
     * @param cursor The pagination cursor.
     * @return A response containing the children of the blog.
     */
    public ChildrenResponse getBlogChildren(long blogId, long cursor) {
        User user = getAuthenticatedUser();
        Pageable pageable = PageRequest.of(0, 10, Sort.by("createdAt", "id").descending());
        Page<Blog> page;
        if (cursor == 0) {
            cursor = Long.MAX_VALUE;
        }
        page = this.blogRepository.findByParentIdAndIdLessThanAndHiddenFalse(blogId, cursor, pageable);

        ChildrenResponse children = new ChildrenResponse();
        children.setChildren(page.stream().map(blog -> new BlogResponse(blog, false, user.getId())).toList());
        children.setCount(this.blogRepository.countByParentId(blogId));
        return children;
    }

    /**
     * Updates an existing blog post.
     * @param blogRequest The request body containing updated blog details.
     * @param blogId The ID of the blog to update.
     * @return The updated blog post.
     */
    @Transactional
    public BlogResponse updateBlog(BlogRequest blogRequest, long blogId) {
        Blog blog = this.blogRepository.findById(blogId).get();
        blog.setDescription(blogRequest.description());
        blog.setTitle(blogRequest.title());
        blog.setUpdatedAt(LocalDateTime.now());
        return new BlogResponse(this.blogRepository.save(blog), blog.getUser().getId());
    }

    /**
     * Hides a blog post.
     * @param blogId The ID of the blog to hide.
     * @return The new hidden status of the blog.
     */
    public boolean hideBlog(long blogId) {
        Blog blog = this.blogRepository.findById(blogId).get();
        blog.setHidden(!blog.isHidden());
        this.blogRepository.save(blog);
        return blog.isHidden();
    }

    /**
     * Converts a BlogRequest DTO to a Blog entity.
     * @param blogRequest The BlogRequest DTO.
     * @return The converted Blog entity.
     */
    public Blog convertToEntity(BlogRequest blogRequest) {
        Blog blog = new Blog();
        if (blogRequest.parent() != null) {
            Blog parent = this.blogRepository.findById(blogRequest.parent())
                    .orElseThrow(() -> new IllegalArgumentException("FAILED: there is no post with this ID"));
            blog.setParent(parent);
        } else {
            if (blogRequest.title() == null || blogRequest.title().trim() == "") {
                throw new IllegalArgumentException("must be to set the title");
            }
            blog.setTitle(blogRequest.title());
        }
        blog.setCreatedAt(LocalDateTime.now());
        blog.setDescription(blogRequest.description());
        return blog;
    }

    /**
     * Updates a Blog entity from a BlogRequest DTO.
     * @param blog The Blog entity to update.
     * @param blogRequest The BlogRequest DTO containing the new data.
     */
    private void updateBlogEntity(Blog blog, BlogRequest blogRequest) {
        blog.setDescription(blogRequest.description());
        blog.setTitle(blogRequest.title());
        blog.setUpdatedAt(LocalDateTime.now());
    }

    /**
     * Toggles a 'like' on a blog post.
     * @param blogId The ID of the blog to like or unlike.
     * @return A map containing the change in the number of likes (1 for like, -1 for unlike).
     */
    public Map<String, Integer> likeBlog(long blogId) {
        User user = getAuthenticatedUser();
        Blog blog = this.blogRepository.findById(blogId).get();
        Like like = this.likesRepository.findByBlogIdAndUserId(blogId, user.getId()).orElse(null);
        if (like == null) {
            like = new Like();
            like.setBlog(blog);
            like.setUser(user);
            this.likesRepository.save(like);
            return Map.of("like", 1);
        } else {
            System.out.println(like.getId());
            this.likesRepository.delete(like);
            return Map.of("like", -1);
        }
    }

    /**
     * Retrieves the total number of likes for a blog post.
     * @param blogId The ID of the blog.
     * @return A map containing the total number of likes.
     */
    public Map<String, Integer> getLikes(long blogId) {
        Blog blog = this.blogRepository.findById(blogId).get();
        return Map.of("count", blog.getLikes().size());
    }

}

File: frontend/src/app/features/blog/services/blog-service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { environment } from '../../../../environments/environment';
import { BlogRequest, BlogResponce, Report } from '../model/model';

/**
 * Service for interacting with the blog API.
 * Provides methods for creating, retrieving, updating, and deleting blog posts,
 * as well as handling likes, reports, and file uploads.
 */
@Injectable({
	providedIn: 'root'
})
export class BlogService {
	private API_URL = environment.API_URL;
    private http = inject(HttpClient)

	/**
	 * Saves a new blog post to the server.
	 * @param blogRequest The blog post data to save.
	 * @returns An Observable of the created BlogResponce.
	 */
	saveBlog(blogRequest: BlogRequest) {
		return this.http.post<BlogResponce>(`${this.API_URL}/blogs/store`, blogRequest)
	}

	/**
	 * Updates an existing blog post on the server.
	 * @param blogRequest The updated blog post data.
	 * @param blogId The ID of the blog post to update.
	 * @returns An Observable of the updated BlogResponce.
	 */
	updateBlog(blogRequest: BlogRequest, blogId: string) {
		return this.http.put<BlogResponce>(`${this.API_URL}/blogs/${blogId}/update`, blogRequest)
	}

	/**
	 * Retrieves a list of blog posts with pagination.
	 * @param cursor The ID of the last blog retrieved, used for pagination.
	 * @returns An Observable of an array of BlogResponce.
	 */
	getBlogs(cursor: number) {
		return this.http.get<BlogResponce[]>(`${this.API_URL}/blogs?cursor=${cursor}`)
	}

	/**
	 * Retrieves blog posts from the user's network with pagination.
	 * @param cursor The ID of the last blog retrieved, used for pagination.
	 * @returns An Observable of an array of BlogResponce.
	 */
	getBlogsNetworks(cursor: number) {
		return this.http.get<BlogResponce[]>(`${this.API_URL}/blogs/networks?cursor=${cursor}`)
	}

	/**
	 * Retrieves blog posts by a specific user with pagination.
	 * @param userId The ID of the user whose blogs to retrieve.
	 * @param cursor The ID of the last blog retrieved, used for pagination.
	 * @returns An Observable of an array of BlogResponce.
	 */
	getBlogsByUserId(userId: any, cursor: number) {
		return this.http.get<BlogResponce[]>(`${this.API_URL}/users/${userId}/blogs?cursor=${cursor}`)
	}

	/**
	 * Retrieves a single blog post by its ID.
	 * @param blogId The ID of the blog post to retrieve.
	 * @returns An Observable of the BlogResponce.
	 */
	getBlog(blogId: any) {
		return this.http.get<BlogResponce>(`${this.API_URL}/blogs/${blogId}`)
	}

	/**
	 * Retrieves children (comments/replies) for a given blog post with pagination.
	 * @param blogId The ID of the parent blog post.
	 * @param cursor The ID of the last child retrieved, used for pagination.
	 * @returns An Observable containing an array of BlogResponce children and their count.
	 */
	getBlogChildren(blogId: any, cursor: number) {
		return this.http.get<{ children: BlogResponce[], count: number }>(`${this.API_URL}/blogs/${blogId}/children?cursor=${cursor}`)
	}

	/**
	 * Deletes a blog post by its ID.
	 * @param blogId The ID of the blog post to delete.
	 * @returns An Observable that completes when the deletion is successful.
	 */
	DeleteBlog(blogId: number) {
		return this.http.delete<void>(`${this.API_URL}/blogs/${blogId}/delete`, { responseType: 'text' as 'json' })
	}

	/**
	 * Hides a blog post by its ID.
	 * @param blogId The ID of the blog post to hide.
	 * @returns An Observable that completes when the hide operation is successful.
	 */
	HideBlog(blogId: number) {
		return this.http.patch<void>(`${this.API_URL}/blogs/${blogId}/hide`, { responseType: 'text' as 'json' })
	}

	/**
	 * Submits a report for a blog post.
	 * @param report The report object containing blogId and reason.
	 * @returns An Observable of the created Report.
	 */
	Report(report: any) {
		return this.http.post<Report>(`${this.API_URL}/reports/store`, report, { responseType: 'text' as 'json' })
	}

	/**
	 * Toggles the like status for a blog post.
	 * @param blogResponce The blog post for which to toggle the like.
	 * @returns An Observable indicating the like status change.
	 */
	toggleLike(blogResponce: BlogResponce) {
		return this.http.post<{ like: number }>(`${this.API_URL}/blogs/${blogResponce.id}/like`, {})
	}

	/**
	 * Uploads a file to the server.
	 * @param file The file to upload.
	 * @returns An Observable containing the URL of the uploaded file.
	 */
	uploadFile(file: File) {
		const formData = new FormData();
		formData.append('file', file)
		return this.http.post<{ url: string }>(`${this.API_URL}/upload`, formData)
	}

	// getLikes(blogResponce: BlogResponce) {
	// 	return this.http.get<{count: number}>(`${this.API_URL}/blogs/${blogResponce.id}/likes`, {})
	// }
}
